cc/algorithms/partition-selection-strategies.h

```diff
@@ -0,0 +1,345 @@
+       return magic;
+     }
+   }
+   protected:
+     base::Status EpsilonIsSetAndValid() {
+       if (!epsilon_.has_value()) {
+         return base::InvalidArgumentError("Epsilon has to be set.");
+       }
+       if (!std::isfinite(epsilon_.value())) {
+         return base::InvalidArgumentError(
+           absl::StrCat("Epsilon has to be finite but is ", epsilon_.value()));
+       }
+       if (epsilon_.value() <= 0) {
+         return base::InvalidArgumentError(
+           absl::StrCat("Epsilon has to be positive but is ",
+                        epsilon_.value()));
+       }
+       return base::OkStatus();
+     }
+  };
+
+  PreaggPartitionSelection(double epsilon, double delta, int max_partitions)
+  	: PartitionSelectionStrategy(epsilon, delta, max_partitions) {
+    if(epsilon_ != 0) {
+      adjusted_epsilon_ = epsilon / (double) max_partitions;
+    }
+    else {
+      adjusted_epsilon_ = 0;
+    }
+    crossover_1_ = 1 + floor((1.0/adjusted_epsilon_)
+                     * log((exp(adjusted_epsilon_)
+                     + 2.0 * adjusted_delta_ - 1.0)
+                       / ((exp(adjusted_epsilon_) + 1) * adjusted_delta_)));
+    crossover_2_ = crossover_1_ + floor((1.0/adjusted_epsilon_)
+                                  *  log(1 + ((exp(adjusted_epsilon_) - 1)
+                                  / adjusted_delta_)
+                                  * (1 - ProbabilityOfKeep(crossover_1_))));
+  }
+
+  virtual ~PreaggPartitionSelection() = default;
+
+  double GetAdjustedEpsilon() { return adjusted_epsilon_; }
+
+  double GetFirstCrossover() { return crossover_1_; }
+
+  double GetSecondCrossover() { return crossover_2_; }
+
+  bool ShouldKeep(int num_users) override {
+    //generate a random number between 0 and 1
+    double rand_num = UniformDouble();
+    //only keep partition if random number < expected probability of keep
+    return (rand_num <= ProbabilityOfKeep(num_users));
+  }
+
+ protected:
+  double adjusted_epsilon_;
+  double crossover_1_;
+  double crossover_2_;
+
+  double ProbabilityOfKeep(double n) {
+    if(n == 0) {
+      return 0;
+    }
+    else if (n <= crossover_1_) {
+      return (((exp(n * adjusted_epsilon_) - 1) / (exp(adjusted_epsilon_) - 1))
+              * adjusted_delta_);
+    }
+    else if (n > crossover_1_ && n <= crossover_2_) {
+      double m = n - crossover_1_;
+      return ((1 - exp(-1 * m * adjusted_epsilon_))
+               * (1 + adjusted_delta_ / (exp(adjusted_epsilon_) - 1))
+               + exp(-1 * m * adjusted_epsilon_)
+               * ProbabilityOfKeep(crossover_1_));
+    }
+    else {
+      return 1;
+    }
+  }
+
+};
+
+
+class LaplacePartitionSelection : public PartitionSelectionStrategy {
+ public:
+  //Builder for LaplacePartitionSelection
+  class Builder : public PartitionSelectionStrategy::Builder {
+   public:
+     Builder& SetLaplaceMechanism(LaplaceMechanism::Builder* laplace_builder) {
```

Comment:
Do you require laplace or could you also use the more generic `NumericalMechanism::Builder`?

Reply:
I guess I could use a generic NumericalMechanism::Builder?  At least that shouldn't impact the code as it currently is.  But, I think the rationale behind using a Laplace builder instead is that the LaplacePartitionSelection class will calculate a threshold based on the CDF of Laplace, so accepting any type of mechanism that might not be sampling from the Laplace distribution shouldn't happen, which is why the type of the argument is set to a LaplaceMechanism::Builder.